sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -391,6 +403,123 @@ DataSource buildDatasource() {
     void setParameters(PreparedStatement preparedStatement) throws Exception;
   }
 
+  /** Implementation of {@link #readRows()}. */
+  @AutoValue
+  @Experimental(Experimental.Kind.SCHEMAS)
+  public abstract static class ReadRows extends PTransform<PBegin, PCollection<Row>> {
```

Comment:
can we share the part that's common with Read? I'm worried that that people will end up adding features to one codepath and not the other.

Reply:
Do you have any suggestions on how to do this? I wasn't happy with the duplication either but I couldn't see a nice way to share state between AutoValue classes. Both the existing `Read` and `ReadAll` classes are pretty much identical to each other as well so I assumed that was a known issue.